### DISCUSSION TITLE
Remapping missing blocks/items/entities

### DISCUSSION THREAD

**Discussion by Zifiv** [FabricMC/fabric-api#2361] [Mod Dev Support] [answered]
Hello all,

I've searched for an answer about this subject in the wiki, but I saw nothing than can help me.

**How can I remap missing blocks/items/entities due to a rename of them in my mod ?**

Thanks for your help, and It could be very usefull to have a wiki page for it.

**Comment by warjort** [comment]
The official way to do this is to use Mojang's DataFixerUpper.
But that has long standing issue about lack of documentation/examples: https://github.com/Mojang/DataFixerUpper/issues/19

Somebody did start some wiki pages for it on the fabric wiki, but they are not complete: https://fabricmc.net/wiki/tutorial:datafixer

Your best bet is probably to look at what Minecraft does itself, e.g. ````net.minecraft.datafixer.Schemas```` uses a BlockNameFix, and ItemNameFix amongst many other things.
e.g. Minecraft's rename of sea_grass to seagrass
````
        Schema schema57 = builder.addSchema(1484, EMPTY_IDENTIFIER_NORMALIZE);
        builder.addFixer(ItemNameFix.create(schema57, "Rename seagrass items", Schemas.replacing(ImmutableMap.of("minecraft:sea_grass", "minecraft:seagrass", "minecraft:tall_sea_grass", "minecraft:tall_seagrass"))));
        builder.addFixer(BlockNameFix.create(schema57, "Rename seagrass blocks", Schemas.replacing(ImmutableMap.of("minecraft:sea_grass", "minecraft:seagrass", "minecraft:tall_sea_grass", "minecraft:tall_seagrass"))));
````

You might also find an example similar to your use case on github? https://github.com/search?l=Java&q=datafixers&type=Code

**Reply by Zifiv** [reply]
There is not Fabric API Callback or Registry to add somes fixer ?

**Reply by warjort** [reply]
I am not aware of such a thing. 

A search on github reveals somebody starting a mod to add such an api, 
but I don't think it is released and it has no documentation?
https://github.com/gudenau/minecraft-DataFixerApi

Trying to find a mod doing something simple with this stuff is quite hard. :-)
I believe the issue is that the datafixer stuff gets built before mods are loaded so you can't just do stuff in your mod initialiser, it would be too late. The above api solves that by adding a new entrypoint.

Here's an example of a mod adding a fixer by injecting into the end of ````Schemas.build(DataFixerBuilder)````
https://github.com/TerraformersMC/Terraform/blob/97d3b10aa0c8492c87ee84aa341c20fec1532acb/terraform-biome-remapper-api-v1/src/main/java/com/terraformersmc/terraform/biomeremapper/impl/mixin/MixinSchemas.java
This is to provide a biome remapping (fixer) api: https://github.com/TerraformersMC/Terraform/blob/97d3b10aa0c8492c87ee84aa341c20fec1532acb/terraform-biome-remapper-api-v1/src/main/java/com/terraformersmc/terraform/biomeremapper/api/BiomeRemapperApi.java

**Reply by Technici4n** [reply]
DFU is not suitable for mods afaik. The "correct" way to handle this is to just wait until the next major MC version to change item ids, or temporarily support the old item with a migration path (think "disabled" machine with a recipe to turn it into the new version of the machine).

**Reply by Zifiv** [reply]
@Technici4n You seems to not understand what I want, I've changed the id of one of my block and block entity that I've added to my mod, and I want to have the possibility to "migrate" the id for people that have created world with older version of my mod to migrate it to the new id.

@warjort Ok then, to inject my DataFixer I need to use a mixin to the `Schemas.class`. Ok, I'll try it, and come back here to feedback about it :) ! Thanks for your help.

Btw, what's about https://github.com/FabricMC/fabric/tree/1.19/fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry, there is some callback about remaping.

**Reply by Technici4n** [reply]
I am saying that unfortunately changing ids is generally not the way to go.

**Reply by warjort** [reply]
> Btw, what's about https://github.com/FabricMC/fabric/tree/1.19/fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry, there is some callback about remaping.

That is for remapping a different id, the raw ids so they agree between client and server. 
Every registry entry has an Identifier (e.g. "minecraft:stone") and a raw id - a unique int.

**Reply by Juuxel** [reply]
Data fixers aren't generally usable for mods. They only run when the Minecraft data version/schema version changes (= on MC version updates), so they won't run if your mod updates and the MC version doesn't. This includes injecting to the Schemas class.

If you create a new schema there, it has a high chance of colliding with a future MC schema version and breaking the worlds from actually upgrading as they should!

This would be fixed by the various data fixer PRs we've had for Fabric API since they track mod schema versions too, but they've become abandoned and Fabric API maintainers also seem to be opposed to including it.

**Reply by Zifiv** [reply]
So it's impossible for modders to having an automatic migration system while we change a block/items/entities id ?

So like @Technici4n said, my only possible way is to have both version of the block, with the old id and with the new one; and only the new id can be crafted or obtainable...

Just to be sure that everyone understand me, in forge there is an event that can be used to remap blocks, items and entites id created by mod : `RegistryEvent.MissingMappings.Mapping`.

**Reply by Juuxel** [reply]
Fabric API doesn't have anything for replacing missing IDs currently, no. A solution using DataFixerUpper is sadly unlikely, but there could be something like Forge's event in the future.

A hacky "solution" that some modders have used is to mixin to `Identifier.<init>` and replace the old components with the new ones there, but that makes it impossible to actually construct those IDs for using them elsewhere so I wouldn't recommend it.

**Reply by Zifiv** [reply]
Ok big thanks to all of u (@warjort, @Technici4n and @Juuxel).

So in Fabric, modders cannot change blocks, items or entities id without break world data of players... it's sad because isn't like it's never arrived for modders to change ids...

**Reply by warjort** [reply]
> A hacky "solution" that some modders have used is to mixin to Identifier.<init> and replace the old components with the new ones there, but that makes it impossible to actually construct those IDs for using them elsewhere so I wouldn't recommend it.

A less hacky fix is to change the ids as they are loaded from the nbt using mixins in 
````net.minecraft.nbt.NbtHelper.toBlockState()```` and ````ItemStack.fromNbt()````
For blocks the tag name is "Name" while for item stacks it is "id".

e.g. something like:
````
@Mixin(NbtHelper.class)
public class NbtHelperMixin {

    @Inject(method = "toBlockState", at = @At("HEAD"), cancellable=true)
    private static void fix_toBlockState(CompoundTag tag, CallbackInfoReturnable<BlockState> callbackInfo) {
        if (tag.contains("Name", NbtElement.STRING_TYPE)) {
            String id = tag.getString("Name");
            if (id.equals(OLD_NAME)) {
                tag.put("Name", NEW_NAME);
            }
        }
        // fall through into the real method with the nbt modified
   }
````

**Reply by warjort** [reply]
The equivalent for an entity would probably be ````EntityType.fromNBT()````, it uses an "id" tag.

**Comment by Zifiv** [comment]
> > A hacky "solution" that some modders have used is to mixin to Identifier. and replace the old components with the new ones there, but that makes it impossible to actually construct those IDs for using them elsewhere so I wouldn't recommend it.
> 
> A less hacky fix is to change the ids as they are loaded from the nbt using mixins in `net.minecraft.nbt.NbtHelper.toBlockState()` and `ItemStack.fromNbt()` For blocks the tag name is "Name" while for item stacks it is "id".
> 
> e.g. something like:
> 
> ```
> @Mixin(NbtHelper.class)
> public class NbtHelperMixin {
> 
>     @Inject(method = "toBlockState", at = @At("HEAD"), cancellable=true)
>     private static void fix_toBlockState(CompoundTag tag, CallbackInfoReturnable<BlockState> callbackInfo) {
>         if (tag.contains("Name", NbtElement.STRING_TYPE)) {
>             String id = tag.getString("Name");
>             if (id.equals(OLD_NAME)) {
>                 tag.put("Name", NEW_NAME);
>             }
>         }
>         // fall through into the real method with the nbt modified
>    }
> ```


So, thanks you @warjort, it worked well for ItemStack and BlockEntity (for BlockEntity, I've created a mixin in `BlockEntity` class and inject at head of the method `createFromNbt`), but for block, it doesn't work.

The world loader seems not to use `toBlockState` from `NbtHelper` class, so my block is just replaced by air.
For now my solution is to keep the old block name.

**Comment by warjort** [comment, marked as answer]
Yes, the latest minecraft doesn't use NbtHelper for loading block states anymore, it does still use it in some other places.

Now it uses the BlockState.CODEC

To make that work, you would need to replace that codec with a modified version of Registry.BLOCK.codec() that includes your mapping.

It is this part you would need to change, that takes an Identifier and looks for a block in the registry.
````id -> Optional.ofNullable(this.get((Identifier)id))````

**Reply by Zifiv** [reply]
Thanks for your reply, but I don't know how to do that with mixins.

I've finally used the hacky solution mentioned by @Juuxel (only for BlockState) that consist on modify the parameter `String[]` of the `Identifier` class constructor to replace old id by new.

I do not like this solution but, it's working well 😃.

**Reply by warjort** [reply]
https://fabricmc.net/wiki/tutorial:mixin_accessors

You would need to mark your accessor as @Mutable since the field is final

Here's an example accessing final static fields:
https://github.com/FabricMC/fabric/blob/653223279f55f79898402b8463a03c95ef5f79f0/fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/mixin/registry/sync/AccessorDebugChunkGenerator.java

On modifying the Identifier, its not a general solution. How do you know the Identifier is for a block?
You would be better off having a mixin for DefaultedRegistry.get(), where you can at least check ````this == Registry.BLOCK````

**Reply by Zifiv** [reply]
You right, it's better by mixin the DefaultedRegistry 👍. Thanks you again.

Maybe later I'll rewrite this to use and recreate a custom Codec that include a migration mapping (generic to be used in other mod I create), because this is the really best way to do migration.

**Reply by NebelNidas** [reply]
For anyone reading this thread after 1.19.3 has dropped, you now have to mixin into `SimpleDefaultedRegistry` instead, as `DefaultedRegistry` has been turned into an interface. See [MythicMaterials](https://github.com/Noaaan/MythicMetals/blob/34959d8caf773d0c54c81281b6e57b19662579b3/src/main/java/nourl/mythicmetals/mixin/DefaultedRegistryMixin.java) for example
